[#48096] Create a storage command to create a (project) folder

ba1ash opened the pull request:
https://community.openproject.org/work_packages/48096

ba1ash committed:
[#48096] Create a storage command to create a (project) folder

https://community.openproject.org/work_packages/48096

apfohl commented on modules/storages/app/common/storages/peripherals/storage_interaction/nextcloud/set_permissions_command.rb:
```diff
@@ -78,24 +87,6 @@ def requested_folder(folder)
-      end
-    end
-
-    def error_result(code, log_message = nil, data = nil)
-      ServiceResult.failure(
-        result: code, # This is needed to work with the ConnectionManager token refresh mechanism.
-        errors: Storages::StorageError.new(code:, log_message:, data:)
-      )
-    end
```
Where did it go?

ba1ash replied:
We have the same method, but with different name `error` inside `Errors` module. So, this one is removed in favor for other one from the module.

apfohl commented on modules/storages/app/common/storages/peripherals/storage_interaction/nextcloud/create_folder_command.rb:
```diff
@@ -0,0 +1,95 @@
+      Nokogiri::XML(response.body).xpath("//s:message").text
+    end
+
+    def escape_whitespace(value)
+      value.gsub(' ', '%20')
+    end
```
Maybe we can use the opportunity to factor out these simple methods to common location.

ba1ash replied:
Good idea, I will do it.

apfohl commented on modules/storages/app/controllers/storages/admin/projects_storages_controller.rb:
```diff
@@ -149,28 +149,37 @@ def set_permissions
-        on_success: ->(_) { flash[:notice] = 'Permissions were successfuly updated on the NextCloud side' }, # rubocop:disable Rails/I18nLocaleTexts
-        on_failure: ->(error) { flash[:error] = "Error: #{error}" }
+      create_folder_command = Storages::Peripherals::StorageRequests.new(storage:).create_folder_command
+      create_folder_command.result.call(folder:).match(
```
Instead of nesting another call returning a `ServiceResult` inside the "success" case of a `.match()`. Consider extracting the inner call to a method and build a pipe line with the `>>` (bind) operator from the ServiceResult refinements. It's exactly made for these cases. See `UploadLinkQuery`.

ba1ash replied:
I am not sure about the usage of this construction in this case. Okay, I will give it a try.

ba1ash committed:
Introduce new module Util inside storages.

The idea is to place some static methods inside it and use them
without need to include a whole module to the class.
